indexing-service/src/main/java/org/apache/druid/indexing/worker/IntermediaryDataManager.java

```diff
@@ -0,0 +1,310 @@
+                }
+              }
+            }
+          }
+          catch (Exception e) {
+            log.warn(e, "Error while cleaning up partitions for expired supervisors");
+          }
+        },
+        intermediaryPartitionCleanupPeriodSec,
+        intermediaryPartitionCleanupPeriodSec,
+        TimeUnit.SECONDS
+    );
+  }
+
+  @LifecycleStop
+  public void stop() throws InterruptedException
+  {
+    if (supervisorTaskChecker != null) {
+      supervisorTaskChecker.shutdownNow();
+      supervisorTaskChecker.awaitTermination(10, TimeUnit.SECONDS);
+    }
+    supervisorTaskCheckTimes.clear();
+  }
+
+  /**
+   * Write a segment into one of configured locations. The location to write is chosen in a round-robin manner per
+   * supervisorTaskId.
+   *
+   * This method is only useful for the new Indexer model. Tasks running in the existing middleManager should the static
+   * addSegment method.
+   */
+  public void addSegment(String supervisorTaskId, String subTaskId, DataSegment segment, File segmentFile)
+      throws IOException
+  {
+    final Iterator<StorageLocation> iterator = locationIterators.computeIfAbsent(
+        supervisorTaskId,
+        k -> Iterators.cycle(shuffleDataLocations)
+    );
+    addSegment(iterator, shuffleDataLocations.size(), supervisorTaskId, subTaskId, segment, segmentFile);
+  }
+
+  /**
+   * Iterate through the given storage locations to find one which can handle the given segment.
+   */
+  public static void addSegment(
+      Iterator<StorageLocation> cyclicIterator,
+      int numLocations,
+      String supervisorTaskId,
+      String subTaskId,
+      DataSegment segment,
+      File segmentFile
+  ) throws IOException
+  {
+    final StorageLocation location = findLocationForSegment(cyclicIterator, numLocations, segment);
+    final File destFile = new File(
+        getPartitionDir(location, supervisorTaskId, segment.getInterval(), segment.getShardSpec().getPartitionNum()),
+        subTaskId
+    );
+    FileUtils.forceMkdirParent(destFile);
+    final long copiedBytes = Files.asByteSource(segmentFile).copyTo(Files.asByteSink(destFile));
+    if (copiedBytes == 0) {
+      throw new IOE(
+          "0 bytes copied after copying a segment file from [%s] to [%s]",
+          segmentFile.getAbsolutePath(),
+          destFile.getAbsolutePath()
+      );
+    }
+    location.addFile(destFile);
+  }
+
+  private static StorageLocation findLocationForSegment(
+      Iterator<StorageLocation> cyclicIterator,
+      int numLocations,
+      DataSegment segment
+  )
+  {
+    for (int i = 0; i < numLocations; i++) {
+      final StorageLocation location = cyclicIterator.next();
+      if (location.canHandle(segment)) {
+        return location;
+      }
+    }
+    throw new ISE("Can't find location to handle segment[%s]", segment);
+  }
+
+  public List<File> findPartitionFiles(String supervisorTaskId, Interval interval, int partitionId)
```

Comment:
nit: could you put this and the non-static method that follows up with the other non-static methods?

Reply:
Moved around to put non-static methods together.